### PR TITLE
FEATURE: Implement Flags on Post Voting Comments, Comment edit and delete for moderators, only_allow_post_voting_in_this_category

### PR DESCRIPTION
**Pull request by jdmartinez1062**
**Context**
The Post Voting plugin allows the creation of Post Voting topics and allows users to comment (Post Voting Comments) and vote on the posts posted

**Problem**
There are 3 problems. First, moderators can't edit or delete Post Voting Comments. Second, customers are unable to enforce the creation of Post Voting Topics in a given category. Finally, Post Voting Comments can't be flagged because such a feature isn't implemented yet

**Solution**
In the same manner and order the problems were presented, solutions will be provided. First, to allow editing and deletion of comments by moderators, we need to modify the guardian methods that allow such actions and any related logic in the FE. Second, create a category setting that forces users to only be able to create Post Voting Topics in a given category. Finally, implement a flagging and reviewable system for Post Voting Comments

### CODE REVIEW

File: test/javascripts/acceptance/composer-test.js [hunk 1]
```diff
@@ -22,6 +22,13 @@ acceptance("Discourse Post Voting - composer", function (needs) {
         createAsPostVotingSetInRequest = true;
       }
 
+      if (
+        parsePostData(request.requestBody).only_post_voting_in_this_category ===
+        "true"
+      ) {
+        createAsPostVotingSetInRequest = true;
+      }
+
```
Comment (janzenisaac) [inline test/javascripts/acceptance/composer-test.js:31]: You could combine this with the condition above

```suggestion
const parsedData = parsePostData(request.requestBody);
if (
  parsedData.create_as_post_voting === "true" ||
  parsedData.only_post_voting_in_this_category === "true"
) {
  createAsPostVotingSetInRequest = true;
}
```

File: spec/models/reviewable_post_voting_comment_spec.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  fab!(:moderator)
+  fab!(:user)
+
+  fab!(:reviewable) do
+    Fabricate(:reviewable_post_voting_comment, target: comment, created_by: moderator)
+  end
+
+  #   it { is_expected.to validate_length_of(:type).is_at_most(100) }
+  #   it { is_expected.to validate_length_of(:target_type).is_at_most(100) }
```
Comment (janzenisaac) [inline spec/models/reviewable_post_voting_comment_spec.rb:23]:
```suggestion
```

File: assets/javascripts/discourse/components/reviewable-post-voting-comment.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import Component from "@glimmer/component";
+import { tracked } from "@glimmer/tracking";
+import { inject as service } from "@ember/service";
+
+export default class ReviewableChatMessage extends Component {
```
Comment (janzenisaac) [inline assets/javascripts/discourse/components/reviewable-post-voting-comment.js:5]: Is this supposed to be `ReviewableChatMessage`? 🚨

File: assets/javascripts/discourse/components/reviewable-post-voting-comment.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  @service store;
+  @tracked post;
+
+  constructor() {
+    super(...arguments);
+    this.store.find("post", this.args.reviewable.post_id).then((post) => {
+      this.post = post;
+    });
+  }
```
Comment (janzenisaac) [inline assets/javascripts/discourse/components/reviewable-post-voting-comment.js:14]:
```suggestion
  get post() {
    return this.store.find("post", this.args.reviewable.post_id).
  };
```

File: lib/post_voting/comment_review_queue.rb [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  class CommentReviewQueue
+    def flag_comment(comment, guardian, flag_type_id, opts = {})
+      result = { success: false, errors: [] }
+      is_notify_user_type = ReviewableScore.types.slice(:notify_user).values.include?(flag_type_id)
+      is_notify_moderators_type =
+        ReviewableScore.types.slice(:notify_moderators).values.include?(flag_type_id)
+      is_notify_type = is_notify_user_type || is_notify_moderators_type
+
+      guardian.ensure_can_flag_post_voting_comment!(comment)
+      guardian.ensure_can_flag_post_voting_comment_as!(comment, flag_type_id, opts)
```
Comment (janzenisaac) [inline lib/post_voting/comment_review_queue.rb:12]: This seems odd, is there any way we could condense this logic, or maybe implement some better naming?

File: lib/post_voting/comment_review_queue.rb [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    def enforce_auto_silence_threshold(reviewable)
+      auto_silence_duration = SiteSetting.chat_auto_silence_from_flags_duration
+      return if auto_silence_duration.zero?
+      return if reviewable.score <= Chat::ReviewableMessage.score_to_silence_user
+
+      user = reviewable.target_created_by
+      return if user.admin?
+      return unless user
+      return if user.silenced?
```
Comment (janzenisaac) [inline lib/post_voting/comment_review_queue.rb:79]: Maybe it should be something like: 

```suggestion
      return unless user
      return if user.admin?
      return if user.silenced?
```

File: assets/javascripts/discourse/lib/post-voting-flag.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    });
+
+    const notifyUserIndex = flagsAvailable.indexOf(
+      flagsAvailable.filterBy("name_key", "notify_user")[0]
+    );
+
+    if (notifyUserIndex !== -1) {
+      const notifyUser = flagsAvailable[notifyUserIndex];
+      flagsAvailable.splice(notifyUserIndex, 1);
+      flagsAvailable.splice(0, 0, notifyUser);
```
Comment (janzenisaac) [inline assets/javascripts/discourse/lib/post-voting-flag.js:38]: ⛳ - maybe

```suggestion
     const notifyUser = flagsAvailable.splice(notifyUserIndex, 1)[0];
     flagsAvailable.unshift(notifyUser);
```

File: assets/javascripts/discourse/components/post-voting-comment-actions.hbs [hunk 8]
```diff
@@ -1,14 +1,25 @@
...
+      @display="link"
+      @class="post-voting-comment-actions-delete-link"
+      @action={{this.deleteConfirm}}
+      @icon="far-trash-alt"
+    />
+
+    {{#if this.canFlag}}
+      <DButton
+        @display="link"
+        @class="post-voting-comment-actions-flag-link"
```
Comment (janzenisaac) [inline assets/javascripts/discourse/components/post-voting-comment-actions.hbs:19]:
```suggestion
        class="post-voting-comment-actions-flag-link"
```

File: assets/javascripts/discourse/components/post-voting-comment-actions.hbs [hunk 9]
```diff
@@ -1,14 +1,25 @@
...
+  <span class="post-voting-comment-actions">
+    <DButton
+      @display="link"
+      @class="post-voting-comment-actions-edit-link"
+      @action={{@updateComment}}
+      @icon="pencil-alt"
+    />
+    <DButton
+      @display="link"
+      @class="post-voting-comment-actions-delete-link"
```
Comment (janzenisaac) [inline assets/javascripts/discourse/components/post-voting-comment-actions.hbs:11]:
```suggestion
      class="post-voting-comment-actions-delete-link"
```

File: assets/javascripts/discourse/components/post-voting-comment-actions.hbs [hunk 10]
```diff
@@ -1,14 +1,25 @@
...
-    @action={{this.deleteConfirm}}
-    @icon="far-trash-alt"
-  />
-</span>
\ No newline at end of file
+{{#if this.canEdit}}
+  <span class="post-voting-comment-actions">
+    <DButton
+      @display="link"
+      @class="post-voting-comment-actions-edit-link"
```
Comment (janzenisaac) [inline assets/javascripts/discourse/components/post-voting-comment-actions.hbs:5]:
```suggestion
      class="post-voting-comment-actions-edit-link"
```

File: assets/javascripts/discourse/components/post-voting-comment-actions.js [hunk 11]
```diff
@@ -1,12 +1,42 @@
...
+    return (
+      this.currentUser &&
+      (this.comment.user_id === this.currentUser.id ||
+        this.currentUser.admin ||
+        this.currentUser.moderator ||
+        this.currentUser.trust_level >=
+          this.siteSettings.min_trust_to_flag_posts_voting_comments) &&
+      !this.args.disabled
+    );
+  }
```
Comment (janzenisaac) [inline assets/javascripts/discourse/components/post-voting-comment-actions.js:39]: ⛳ There is a lot of shared logic here, can we try to combine some of it?

File: assets/javascripts/discourse/components/post-voting-comment-actions.js [hunk 12]
```diff
@@ -26,4 +56,17 @@ export default class PostVotingCommentActions extends Component {
...
+
+  @action
+  showFlag() {
+    this.comment.availableFlags = this.comment.available_flags;
+    this.modal.show(FlagModal, {
+      model: {
+        flagTarget: new PostVotingFlag(),
+        flagModel: this.comment,
+        setHidden: () => this.comment.set("hidden", true),
+        site: this.site,
```
Comment (janzenisaac) [inline assets/javascripts/discourse/components/post-voting-comment-actions.js:68]:
```suggestion
```

I would import `site` in the modal directly if you need it.

File: assets/javascripts/discourse/components/post-voting-comment-actions.js [hunk 13]
```diff
@@ -26,4 +56,17 @@ export default class PostVotingCommentActions extends Component {
       },
     });
   }
+
+  @action
+  showFlag() {
+    this.comment.availableFlags = this.comment.available_flags;
```
Comment (janzenisaac) [inline assets/javascripts/discourse/components/post-voting-comment-actions.js:62]: I don't think this is necessary
```suggestion
```

File: assets/javascripts/discourse/components/post-voting-comment-actions.js [hunk 14]
```diff
@@ -26,4 +56,17 @@ export default class PostVotingCommentActions extends Component {
...
+    this.modal.show(FlagModal, {
+      model: {
+        flagTarget: new PostVotingFlag(),
+        flagModel: this.comment,
+        setHidden: () => this.comment.set("hidden", true),
```
Comment (janzenisaac) [inline assets/javascripts/discourse/components/post-voting-comment-actions.js:67]: We should try to get away from using `set` and instead use a `tracked` property

**Review by janzenisaac — CHANGES_REQUESTED** [review]

File: assets/javascripts/discourse/components/reviewable-post-voting-comment.js (hunk 4 above)
Comment (janzenisaac) [inline assets/javascripts/discourse/components/reviewable-post-voting-comment.js:14]: I think we could do something like

```suggestion
  constructor() {
    super(...arguments);
    this.fetchPost();
  }
  
  async fetchPost() {
     this.post = await this.store.find("post", this.args.reviewable.post_id);
  }
```

File: assets/javascripts/discourse/components/reviewable-post-voting-comment.js (hunk 3 above)
Comment (jdmartinez1062) [reply assets/javascripts/discourse/components/reviewable-post-voting-comment.js:5]: Yup I swear I changed this like 5 times but always got back jajaja

File: assets/javascripts/discourse/components/post-voting-comment-actions.js (hunk 12 above)
Comment (jdmartinez1062) [reply assets/javascripts/discourse/components/post-voting-comment-actions.js:68]: I think the reviewables pipeline uses this, tried to import it in the `post-voting-flag` but it didn't worked

File: assets/javascripts/discourse/components/post-voting-comment-actions.js (hunk 13 above)
Comment (jdmartinez1062) [reply assets/javascripts/discourse/components/post-voting-comment-actions.js:62]: This is related to the reviewables pipeline too, without this it is unable to identify the proper attributeit tries to use `availableFlags` but it only has `available_flags`

**Review by janzenisaac — APPROVED** [review]
Looks great!

**Pull request merged**